sale_by_packaging/models/sale_order_line.py

```diff
@@ -0,0 +1,160 @@
+# Copyright 2020 Camptocamp SA
+# License AGPL-3.0 or later (https://www.gnu.org/licenses/agpl)
+from odoo import _, api, fields, models
+from odoo.exceptions import ValidationError
+
+
+class SaleOrderLine(models.Model):
+
+    _inherit = "sale.order.line"
+
+    def _can_be_sold_error_condition(self):
+        self.ensure_one()
+        return self.product_packaging and not self.product_packaging.can_be_sold
+
+    @api.constrains("product_packaging")
+    def _check_product_packaging_can_be_sold(self):
+        for line in self:
+            if line._can_be_sold_error_condition():
+                raise ValidationError(
+                    _(
+                        "Packaging %s on product %s must be set as 'Can be sold'"
+                        " in order to be used on a sale order."
+                    )
+                    % (line.product_packaging.name, line.product_id.name)
+                )
+
+    @api.onchange("product_packaging")
+    def _onchange_product_packaging(self):
+        if self._can_be_sold_error_condition():
+            return {
+                "warning": {
+                    "title": _("Warning"),
+                    "message": _(
+                        "This product packaging must be set as 'Can be sold' in"
+                        " order to be used on a sale order."
+                    ),
+                },
+            }
+        return super()._onchange_product_packaging()
+
+    @api.constrains("product_id", "product_packaging")
+    def _check_product_packaging_sell_only_by_packaging(self):
+        for line in self:
+            if line.product_id.sell_only_by_packaging and not line.product_packaging:
+                raise ValidationError(
+                    _(
+                        "Product %s can only be sold with a packaging."
+                        % line.product_id.name
+                    )
+                )
+
+    @api.onchange("product_id")
+    def product_id_change(self):
+        res = super().product_id_change()
+        if self.product_id.sell_only_by_packaging:
+            first_packaging = fields.first(self.product_id.packaging_ids)
+            self.update(
+                {
+                    "product_packaging": first_packaging.id,
+                    "product_uom_qty": first_packaging.qty,
```

Comment:
here you could have ZERO as qty. Can we do anything better? Eg: not setting the qty and / or show a warning?

Reply:
I filtered out packagings with a qty of zero and added a note in roadmap :wink: